packages/integration-tests/websocket-client.js

```diff
@@ -0,0 +1,238 @@
+/* eslint-disable @typescript-eslint/no-var-requires */
+const WebSocket = require('ws');
+const { EventEmitter } = require('events');
+
+const NOT_INITIALIZED = new Error('websocket_not_initialized');
+
+const createDeferred = id => {
+    let localResolve = t => () => {};
+    let localReject = e => () => {};
+
+    const promise = new Promise((resolve, reject) => {
+        localResolve = resolve;
+        localReject = reject;
+    });
+
+    return {
+        id,
+        resolve: localResolve,
+        reject: localReject,
+        promise,
+    };
+};
+
+// Making the timeout high because the controller in trezor-user-env
+// must synchronously run actions on emulator and they may take a long time
+// (for example in case of Shamir backup)
+const DEFAULT_TIMEOUT = 5 * 60 * 1000;
+const DEFAULT_PING_TIMEOUT = 50 * 1000;
+
+class Controller extends EventEmitter {
+    constructor(options) {
+        super();
+        this.messageID = 0;
+        this.messages = [];
+        this.subscriptions = [];
+        this.setMaxListeners(Infinity);
+        this.options = options;
+    }
+
+    setConnectionTimeout() {
+        this.clearConnectionTimeout();
+        this.connectionTimeout = setTimeout(
+            this.onTimeout.bind(this),
+            this.options.timeout || DEFAULT_TIMEOUT,
+        );
+    }
+
+    clearConnectionTimeout() {
+        if (this.connectionTimeout) {
+            clearTimeout(this.connectionTimeout);
+            this.connectionTimeout = undefined;
+        }
+    }
+
+    setPingTimeout() {
+        if (this.pingTimeout) {
+            clearTimeout(this.pingTimeout);
+        }
+        this.pingTimeout = setTimeout(
+            this.onPing.bind(this),
+            this.options.pingTimeout || DEFAULT_PING_TIMEOUT,
+        );
+    }
+
+    onTimeout() {
+        const { ws } = this;
+        if (!ws) return;
+        if (ws.listenerCount('open') > 0) {
+            ws.emit('error', 'Websocket timeout');
+            try {
+                ws.close();
+            } catch (error) {
+                // empty
+            }
+        } else {
+            this.messages.forEach(m => m.reject(new Error('websocket_timeout')));
+            ws.close();
+        }
+    }
+
+    async onPing() {
+        // make sure that connection is alive if there are subscriptions
+        if (this.ws && this.isConnected()) {
+            if (this.subscriptions.length > 0 || this.options.keepAlive) {
+                await this.getBlockHash(0);
+            } else {
+                try {
+                    this.ws.close();
+                } catch (error) {
+                    // empty
+                }
+            }
+        }
+    }
+
+    onError() {
+        this.dispose();
+    }
+
+    send(params) {
+        const { ws } = this;
+        if (!ws) throw NOT_INITIALIZED;
+        const id = this.messageID;
+
+        const dfd = createDeferred(id);
```

Comment:
Naming `dfd` is very confusing to read, since it is the response from `createDeferred` what about calling it `deferred` ?